File: .fullsend/skills/e2e-failure-analysis/scripts/download-artifacts.py
#!/usr/bin/env python3
"""Download E2E test artifacts from a prow/gcsweb URL via the public GCS JSON API."""

import gzip
import json
import os
import re
import shutil
import ssl
import sys
import urllib.parse
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from urllib.parse import urlparse

BUCKET = "test-platform-results"
API_URL = f"https://storage.googleapis.com/storage/v1/b/{BUCKET}/o"
DL_URL = f"https://storage.googleapis.com/{BUCKET}"
MAX_WORKERS = 8

EXCLUDE_RE = re.compile(r"\.webm$|/playwright-report/data/|/playwright-report/trace/")
INCLUDE_RE = re.compile(r"/e2e-test-results/.*/trace\.zip$")


def _ssl_ctx():
    for ca in (os.environ.get("SSL_CERT_FILE"), os.environ.get("REQUESTS_CA_BUNDLE"),
               "/etc/openshell-tls/ca-bundle.pem", "/etc/ssl/certs/ca-certificates.crt"):
        if ca and Path(ca).is_file():
            return ssl.create_default_context(cafile=ca)
    return ssl.create_default_context()


CTX = _ssl_ctx()


def parse_url(url):
    path = urlparse(url.rstrip("/")).path
    for pfx in ("/view/gs/", "/gcs/"):
        if path.startswith(pfx):
            path = path[len(pfx):]
            break
    parts = path.split("/")

    if "pr-logs" in parts:
        i = parts.index("pull") + 1
        pr, job, jid = parts[i + 1], parts[i + 2], parts[i + 3]
        sub = job.split("-main-")[-1] if "-main-" in job else "e2e-ocp-helm"
        return {"type": "pr", "pr": pr, "job_id": jid, "subdir": sub,
                "gcs": f"pr-logs/pull/redhat-developer_rhdh-plugin-export-overlays/{pr}/{job}/{jid}"}

    if "/logs/" in "/" + "/".join(parts):
        i = parts.index("logs")
        job, jid = parts[i + 1], parts[i + 2]
        sub = job.split("-main-")[-1] if "-main-" in job else "e2e-ocp-helm-nightly"
        return {"type": "nightly", "job_id": jid, "subdir": sub,
                "gcs": f"logs/{job}/{jid}"}

    return None


def gcs_list(prefix):
    items, token = [], None
    while True:
        p = {"prefix": prefix, "maxResults": "1000"}
        if token:
            p["pageToken"] = token
        with urllib.request.urlopen(f"{API_URL}?{urllib.parse.urlencode(p)}", context=CTX) as r:
            data = json.loads(r.read())
        items.extend(data.get("items", []))
        token = data.get("nextPageToken")
        if not token:
            return items


def download(gcs_prefix, dest):
    dest.mkdir(parents=True, exist_ok=True)

    print("Listing objects...", file=sys.stderr)
    all_items = gcs_list(gcs_prefix + "/")
    keep = [i for i in all_items if INCLUDE_RE.search(i["name"]) or not EXCLUDE_RE.search(i["name"])]
    total_mb = sum(int(i.get("size", 0)) for i in keep) / 1024 / 1024
    print(f"Downloading {len(keep)}/{len(all_items)} files ({total_mb:.1f} MB)...", file=sys.stderr)

    plen = len(gcs_prefix) + 1

    def _dl(item):
        rel = item["name"][plen:]
        local = dest / rel
        local.parent.mkdir(parents=True, exist_ok=True)
        urllib.request.urlretrieve(f"{DL_URL}/{urllib.parse.quote(item['name'], safe='/')}", str(local))

    failed = 0
    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
        futs = {pool.submit(_dl, i): i for i in keep}
        for f in as_completed(futs):
            try:
                f.result()
            except Exception as e:
                failed += 1
                print(f"  FAILED: {futs[f]['name'].split('/')[-1]}: {e}", file=sys.stderr)

    print(f"Done: {len(keep) - failed}/{len(keep)} files.", file=sys.stderr)

    # Decompress gzipped text files (GCS serves them with raw gzip encoding)
    for path in dest.rglob("*"):
        if not path.is_file() or path.suffix in (".zip", ".gz", ".png", ".webm"):
            continue
        with open(path, "rb") as f:
            if f.read(2) != b"\x1f\x8b":
                continue
        tmp = path.with_suffix(path.suffix + ".tmp")
        with gzip.open(path, "rb") as fi, open(tmp, "wb") as fo:
            shutil.copyfileobj(fi, fo)
        tmp.replace(path)


def is_cache_valid(artifacts_dir):
    if not artifacts_dir.exists():
        return False
    has = lambda g: any(artifacts_dir.rglob(g))
    return sum([has("logs/*/pods.txt"), has("results.json"), has("error-context.md")]) >= 2


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 download-artifacts.py <PROW_URL>", file=sys.stderr)
        sys.exit(1)

    info = parse_url(sys.argv[1])
    if not info:
        print(f"ERROR: Could not parse URL: {sys.argv[1]}", file=sys.stderr)
        sys.exit(1)

    base = Path("node_modules/.cache/e2e-artifacts")
    cache_dir = base / info.get("pr", "nightly") / info["job_id"]
    container = "redhat-developer-rhdh-plugin-export-overlays-ocp-helm"
    artifacts_dir = cache_dir / container / "artifacts"

    if is_cache_valid(artifacts_dir):
        print("Artifacts already downloaded, using cache.", file=sys.stderr)
        print(artifacts_dir)
        return

    if cache_dir.exists():
        print("Cache incomplete, re-downloading...", file=sys.stderr)
        shutil.rmtree(cache_dir)

    gcs_src = f"{info['gcs']}/artifacts/{info['subdir']}/{container}"
    download(gcs_src, cache_dir / container)

    if not artifacts_dir.exists():
        print(f"ERROR: Artifacts dir not found: {artifacts_dir}", file=sys.stderr)
        sys.exit(1)

    print("Download complete.", file=sys.stderr)
    print(artifacts_dir)


if __name__ == "__main__":
    main()

File: .fullsend/skills/e2e-failure-analysis/scripts/download-artifacts.ts
#!/usr/bin/env node
import * as fs from "node:fs/promises";
import { createWriteStream } from "node:fs";
import * as path from "node:path";
import { gunzipSync } from "node:zlib";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";

const BUCKET = "test-platform-results";
const API_URL = `https://storage.googleapis.com/storage/v1/b/${BUCKET}/o`;
const DL_URL = `https://storage.googleapis.com/${BUCKET}`;
const MAX_CONCURRENCY = 8;

const EXCLUDE_RE = /\.webm$|\/playwright-report\/data\/|\/playwright-report\/trace\//;
const INCLUDE_RE = /\/e2e-test-results\/.*\/trace\.zip$/;

interface GCSItem {
  name: string;
  size?: string;
}

interface ParsedURL {
  type: string;
  pr?: string;
  job_id: string;
  subdir: string;
  gcs: string;
}

function parseUrl(url: string): ParsedURL | null {
  const p = new URL(url.replace(/\/$/, "")).pathname
    .replace(/^\/view\/gs\//, "")
    .replace(/^\/gcs\//, "");
  const parts = p.split("/");

  if (parts.includes("pr-logs")) {
    const i = parts.indexOf("pull") + 1;
    const [pr, job, jid] = [parts[i + 1], parts[i + 2], parts[i + 3]];
    const sub = job.includes("-main-") ? job.split("-main-").pop()! : "e2e-ocp-helm";
    return {
      type: "pr", pr, job_id: jid, subdir: sub,
      gcs: `pr-logs/pull/redhat-developer_rhdh-plugin-export-overlays/${pr}/${job}/${jid}`,
    };
  }

  if (parts.includes("logs")) {
    const i = parts.indexOf("logs");
    const [job, jid] = [parts[i + 1], parts[i + 2]];
    const sub = job.includes("-main-") ? job.split("-main-").pop()! : "e2e-ocp-helm-nightly";
    return {
      type: "nightly", job_id: jid, subdir: sub,
      gcs: `logs/${job}/${jid}`,
    };
  }

  return null;
}

async function gcsList(prefix: string): Promise<GCSItem[]> {
  const items: GCSItem[] = [];
  let pageToken: string | undefined;
  do {
    const params = new URLSearchParams({ prefix, maxResults: "1000" });
    if (pageToken) params.set("pageToken", pageToken);
    const res = await fetch(`${API_URL}?${params}`);
    if (!res.ok) throw new Error(`GCS list failed: ${res.status} ${res.statusText}`);
    const data = await res.json();
    if (data.items) items.push(...data.items);
    pageToken = data.nextPageToken;
  } while (pageToken);
  return items;
}

async function downloadFile(item: GCSItem, prefixLen: number, dest: string): Promise<void> {
  const local = path.join(dest, item.name.slice(prefixLen));
  await fs.mkdir(path.dirname(local), { recursive: true });
  const res = await fetch(`${DL_URL}/${encodeURIComponent(item.name).replace(/%2F/g, "/")}`);
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  await pipeline(Readable.fromWeb(res.body as any), createWriteStream(local));
}

async function runPool(items: GCSItem[], fn: (item: GCSItem) => Promise<void>): Promise<number> {
  let failed = 0;
  let idx = 0;

  async function worker(): Promise<void> {
    while (idx < items.length) {
      const item = items[idx++];
      try {
        await fn(item);
      } catch (e) {
        failed++;
        process.stderr.write(`  FAILED: ${item.name.split("/").pop()}: ${e}\n`);
      }
    }
  }

  await Promise.all(Array.from({ length: MAX_CONCURRENCY }, worker));
  return failed;
}

async function decompressGzipped(dir: string): Promise<void> {
  const skipExts = new Set([".zip", ".gz", ".png", ".webm"]);
  const gzipMagic = Buffer.from([0x1f, 0x8b]);

  const entries = await fs.readdir(dir, { withFileTypes: true, recursive: true });
  await Promise.all(
    entries
      .filter((e) => e.isFile() && !skipExts.has(path.extname(e.name)))
      .map(async (e) => {
        const full = path.join(e.parentPath || e.path, e.name);
        const fd = await fs.open(full, "r");
        const header = Buffer.alloc(2);
        const { bytesRead } = await fd.read(header, 0, 2, 0);
        await fd.close();
        if (bytesRead < 2 || !header.subarray(0, 2).equals(gzipMagic)) return;
        const decompressed = gunzipSync(await fs.readFile(full));
        await fs.writeFile(full, decompressed);
      }),
  );
}

async function main(): Promise<void> {
  if (process.argv.length < 3) {
    process.stderr.write("Usage: node download-artifacts.ts <PROW_URL>\n");
    process.exit(1);
  }

  const info = parseUrl(process.argv[2]);
  if (!info) {
    process.stderr.write(`ERROR: Could not parse URL: ${process.argv[2]}\n`);
    process.exit(1);
  }

  const base = path.join("node_modules", ".cache", "e2e-artifacts");
  const cacheDir = path.join(base, info.pr || "nightly", info.job_id);
  const container = "redhat-developer-rhdh-plugin-export-overlays-ocp-helm";
  const artifactsDir = path.join(cacheDir, container, "artifacts");

  await fs.rm(cacheDir, { recursive: true, force: true });

  const gcsSrc = `${info.gcs}/artifacts/${info.subdir}/${container}`;

  process.stderr.write("Listing objects...\n");
  const allItems = await gcsList(gcsSrc + "/");
  const keep = allItems.filter((i) => INCLUDE_RE.test(i.name) || !EXCLUDE_RE.test(i.name));
  const totalMb = keep.reduce((s, i) => s + parseInt(i.size || "0", 10), 0) / 1024 / 1024;
  process.stderr.write(`Downloading ${keep.length}/${allItems.length} files (${totalMb.toFixed(1)} MB)...\n`);

  const prefixLen = gcsSrc.length + 1;
  const failed = await runPool(keep, (item) => downloadFile(item, prefixLen, cacheDir + "/" + container));

  process.stderr.write(`Done: ${keep.length - failed}/${keep.length} files.\n`);
  await decompressGzipped(cacheDir);

  if (!(await fs.stat(artifactsDir).catch(() => null))) {
    process.stderr.write(`ERROR: Artifacts dir not found: ${artifactsDir}\n`);
    process.exit(1);
  }

  process.stderr.write("Download complete.\n");
  process.stdout.write(artifactsDir + "\n");
}

main().catch((err) => {
  process.stderr.write(`ERROR: ${err}\n`);
  process.exit(1);
});
